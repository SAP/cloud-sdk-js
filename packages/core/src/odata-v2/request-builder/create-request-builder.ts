import {
  Constructable,
  EntityIdentifiable,
  CreateRequestBuilder as CreateRequestBuilderBase
} from '../../odata-common';
import { EntityV2 } from '../entity';
import { entityDeserializerV2 } from '../entity-deserializer';
import { entitySerializerV2 } from '../entity-serializer';
import { oDataUriV2 } from '../uri-conversion';
import { responseDataAccessorV2 } from './response-data-accessor';
/**
 * Create OData request to create an entity.
 *
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilderV2<EntityT extends EntityV2>
  extends CreateRequestBuilderBase<EntityT>
  implements EntityIdentifiable<EntityT> {
  /**
   * Creates an instance of CreateRequestBuilder.
   *
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param _entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly _entity: EntityT
  ) {
    super(
      _entityConstructor,
      _entity,
      oDataUriV2,
      entitySerializerV2,
      entityDeserializerV2,
      responseDataAccessorV2
    );
  }
}

export { CreateRequestBuilderV2 as CreateRequestBuilder };
