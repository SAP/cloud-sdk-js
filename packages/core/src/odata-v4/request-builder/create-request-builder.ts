import {
  Constructable,
  CreateRequestBuilder as CreateRequestBuilderBase,
  EntityIdentifiable
} from '../../odata-common';
import { Entity } from '../entity';
import { oDataUri } from '../uri-conversion';
import { entitySerializer } from '../entity-serializer';
import { entityDeserializer } from '../entity-deserializer';
import { responseDataAccessor } from './response-data-accessor';
/**
 * Create OData request to create an entity.
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilder<EntityT extends Entity>
  extends CreateRequestBuilderBase<EntityT>
  implements EntityIdentifiable<EntityT>
{
  /**
   * Creates an instance of CreateRequestBuilder.
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param _entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly _entity: EntityT
  ) {
    super(
      _entityConstructor,
      _entity,
      oDataUri,
      entitySerializer,
      entityDeserializer,
      responseDataAccessor
    );
  }
}

export { CreateRequestBuilder as CreateRequestBuilderV4 };
