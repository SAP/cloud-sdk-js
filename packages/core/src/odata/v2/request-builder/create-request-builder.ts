import { errorWithCause } from '@sap-cloud-sdk/util';
import { Constructable, EntityIdentifiable, Link } from '../../common';
import { MethodRequestBuilderBase } from '../../common/request-builder/request-builder-base';
import { ODataCreateRequestConfig } from '../../common/request/odata-create-request-config';
import { EntityV2 } from '../entity';
import { deserializeEntityV2 } from '../entity-deserializer';
import { serializeEntityV2 } from '../entity-serializer';
import { DestinationOptions } from '../../../scp-cf';
import {
  Destination,
  DestinationNameAndJwt
} from '../../../scp-cf/destination-service-types';
import { oDataUriV2 } from '../uri-conversion';
import { getSingleResult } from './response-data-accessor';
/**
 * Create OData request to create an entity.
 *
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilderV2<EntityT extends EntityV2>
  extends MethodRequestBuilderBase<ODataCreateRequestConfig<EntityT>>
  implements EntityIdentifiable<EntityT> {
  /**
   * Creates an instance of CreateRequestBuilder.
   *
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param _entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly _entity: EntityT
  ) {
    super(new ODataCreateRequestConfig(_entityConstructor, oDataUriV2));
    this.requestConfig.payload = serializeEntityV2(
      this._entity,
      this._entityConstructor
    );
  }

  get entity(): EntityT {
    return this._entity;
  }

  /**
   * @deprecated Since v1.29.0. This method should never be called, it has severe side effects.   * Builds the payload of the query.
   * @returns the builder itself
   */
  prepare(): this {
    this.requestConfig.payload = serializeEntityV2(
      this._entity,
      this._entityConstructor
    );
    return this;
  }

  /**
   * Specifies the parent of the entity to create.
   *
   * @param parentEntity - Parent of the entity to create
   * @param linkField - Static representation of the navigation property that navigates from the parent entity to the child entity
   * @returns The entity itself, to facilitate method chaining
   */
  asChildOf<ParentEntityT extends EntityV2>(
    parentEntity: ParentEntityT,
    linkField: Link<ParentEntityT, EntityT>
  ): this {
    this.requestConfig.parentKeys = oDataUriV2.getEntityKeys(
      parentEntity,
      linkField._entityConstructor
    );
    this.requestConfig.childField = linkField;
    return this;
  }

  /**
   * Execute query.
   *
   * @param destination - Destination to execute the request against
   * @param options - Options to employ when fetching destinations
   * @returns A promise resolving to the created entity
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<EntityT> {
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        deserializeEntityV2(
          getSingleResult(response.data),
          this._entityConstructor,
          response.headers
        )
      )
      .catch(error =>
        Promise.reject(errorWithCause('Create request failed!', error))
      );
  }
}

export { CreateRequestBuilderV2 as CreateRequestBuilder };
