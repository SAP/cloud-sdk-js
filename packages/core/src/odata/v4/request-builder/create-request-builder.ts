/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause, MapType } from '@sap-cloud-sdk/util';
import { Constructable, EntityIdentifiable, Link } from '../../common';
import { MethodRequestBuilderBase } from '../../common/request-builder/request-builder-base';
import { ODataCreateRequestConfig } from '../../common/request/odata-create-request-config';
import { Entity } from '../entity';
import { deserializeEntity } from '../entity-deserializer';
import { serializeEntity } from '../entity-serializer';
import { DestinationOptions } from '../../../scp-cf';
import {
  Destination,
  DestinationNameAndJwt
} from '../../../scp-cf/destination-service-types';
import { oDataUri } from '../uri-conversion';
import { HttpReponse } from '../../../http-client';
/**
 * Create OData request to create an entity.
 *
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilder<EntityT extends Entity>
  extends MethodRequestBuilderBase<ODataCreateRequestConfig<EntityT>>
  implements EntityIdentifiable<EntityT> {
  /**
   * Creates an instance of CreateRequestBuilder.
   *
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param _entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly _entity: EntityT
  ) {
    super(new ODataCreateRequestConfig(_entityConstructor, oDataUri));
  }

  get entity(): EntityT {
    return this._entity;
  }

  /**
   * Builds the payload of the query.
   *
   * @returns the builder itself
   */
  prepare(): this {
    this.requestConfig.payload = serializeEntity(
      this._entity,
      this._entityConstructor
    );
    return this;
  }

  /**
   * Specifies the parent of the entity to create.
   *
   * @param parentEntity - Parent of the entity to create
   * @param linkField - Static representation of the navigation property that navigates from the parent entity to the child entity
   * @returns The entity itself, to facilitate method chaining
   */
  asChildOf<ParentEntityT extends Entity>(
    parentEntity: ParentEntityT,
    linkField: Link<ParentEntityT, EntityT>
  ): this {
    this.requestConfig.parentKeys = oDataUri.getEntityKeys(
      parentEntity,
      linkField._entityConstructor
    );
    this.requestConfig.childField = linkField;
    return this;
  }

  /**
   * Execute query.
   *
   * @param destination - Destination to execute the request against
   * @param options - Options to employ when fetching destinations
   * @returns A promise resolving to the created entity
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<EntityT> {
    this.prepare();
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        deserializeEntity(
          extractData(response),
          this._entityConstructor,
          response.headers
        )
      )
      .catch(error =>
        Promise.reject(errorWithCause('Create request failed!', error))
      );
  }
}

function extractData(response: HttpReponse): MapType<any> {
  return response.data;
}
