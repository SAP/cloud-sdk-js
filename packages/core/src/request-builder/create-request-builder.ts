/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { errorWithCause } from '@sap-cloud-sdk/util';
import { Constructable } from '../constructable';
import { Entity, EntityIdentifiable } from '../entity';
import { deserializeEntity } from '../entity-deserializer';
import { serializeEntity } from '../entity-serializer';
import { DestinationOptions } from '../scp-cf';
import {
  Destination,
  DestinationNameAndJwt
} from '../scp-cf/destination-service-types';
import { Link } from '../selectable';
import { MethodRequestBuilderBase } from './request-builder-base';
import { getEntityKeys } from './request/get-keys';
import { ODataCreateRequestConfig } from './request/odata-create-request-config';

/**
 * Create OData request to create an entity.
 *
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilder<EntityT extends Entity>
  extends MethodRequestBuilderBase<ODataCreateRequestConfig<EntityT>>
  implements EntityIdentifiable<EntityT> {
  /**
   * Creates an instance of CreateRequestBuilder.
   *
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly entity: EntityT
  ) {
    super(new ODataCreateRequestConfig(_entityConstructor));
  }

  /**
   * Builds the payload of the query.
   *
   * @returns the builder itself
   */
  prepare(): this {
    this.requestConfig.payload = serializeEntity(
      this.entity,
      this._entityConstructor
    );
    return this;
  }

  /**
   * Specifies the parent of the entity to create.
   *
   * @param parentEntity - Parent of the entity to create
   * @param linkField - Static representation of the navigation property that navigates from the parent entity to the child entity
   * @returns The entity itself, to facilitate method chaining
   */
  asChildOf<ParentEntityT extends Entity>(
    parentEntity: ParentEntityT,
    linkField: Link<ParentEntityT, EntityT>
  ): this {
    this.requestConfig.parentKeys = getEntityKeys(
      parentEntity,
      linkField._entityConstructor
    );
    this.requestConfig.childField = linkField;
    return this;
  }

  /**
   * Execute query.
   *
   * @param destination - Destination to execute the request against
   * @param options - Options to employ when fetching destinations
   * @returns A promise resolving to the created entity
   */
  async execute(
    destination: Destination | DestinationNameAndJwt,
    options?: DestinationOptions
  ): Promise<EntityT> {
    this.prepare();
    return this.build(destination, options)
      .then(request => request.execute())
      .then(response =>
        deserializeEntity(
          response.data.d,
          this._entityConstructor,
          response.headers
        )
      )
      .catch(error =>
        Promise.reject(errorWithCause('Create request failed!', error))
      );
  }
}
