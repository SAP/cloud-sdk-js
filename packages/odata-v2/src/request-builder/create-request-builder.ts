import {
  EntityIdentifiable,
  CreateRequestBuilderBase,
  entitySerializer,
  EntityApi
} from '@sap-cloud-sdk/odata-common/internal';
import {
  DefaultDeSerializers,
  DeSerializers,
  entityDeserializer
} from '../de-serializers';
import { Entity } from '../entity';
import { createODataUri } from '../uri-conversion';
import { responseDataAccessor } from './response-data-accessor';

/**
 * Create OData request to create an entity.
 * @typeParam EntityT - Type of the entity to be created.
 */
export class CreateRequestBuilder<
    EntityT extends Entity,
    DeSerializersT extends DeSerializers = DefaultDeSerializers
  >
  extends CreateRequestBuilderBase<EntityT, DeSerializersT>
  implements EntityIdentifiable<EntityT, DeSerializersT>
{
  /**
   * Creates an instance of CreateRequestBuilder.
   * @param entityApi - Entity API for building and executing the request.
   * @param _entity - Entity to be created.
   */
  constructor(
    entityApi: EntityApi<EntityT, DeSerializersT>,
    readonly _entity: EntityT
  ) {
    super(
      entityApi,
      _entity,
      createODataUri(entityApi.deSerializers),
      entitySerializer(entityApi.deSerializers),
      entityDeserializer(entityApi.deSerializers),
      responseDataAccessor
    );
  }
}
