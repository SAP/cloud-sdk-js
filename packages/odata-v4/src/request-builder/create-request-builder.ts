import {
  Constructable,
  CreateRequestBuilderBase,
  EntityIdentifiable
} from '@sap-cloud-sdk/odata-common/internal';
import { Entity } from '../entity';
import { entitySerializer } from '../entity-serializer';
import { entityDeserializer } from '../entity-deserializer';
import { DeSerializers } from '../de-serializers';
import { responseDataAccessor } from './response-data-accessor';
/**
 * Create OData request to create an entity.
 * @typeparam EntityT - Type of the entity to be created
 */
export class CreateRequestBuilder<
    EntityT extends Entity,
    DeSerializersT extends DeSerializers
  >
  extends CreateRequestBuilderBase<EntityT, DeSerializersT>
  implements EntityIdentifiable<EntityT, DeSerializersT>
{
  /**
   * Creates an instance of CreateRequestBuilder.
   * @param _entityConstructor - Constructor type of the entity to be created
   * @param _entity - Entity to be created
   */
  constructor(
    readonly _entityConstructor: Constructable<EntityT>,
    readonly _entity: EntityT
  ) {
    super(
      _entityConstructor,
      _entity,
      createODataUri(),
      entitySerializer,
      entityDeserializer,
      responseDataAccessor
    );
  }
}
