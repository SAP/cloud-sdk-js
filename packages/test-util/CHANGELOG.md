# @sap-cloud-sdk/test-util

## 2.5.0

### Patch Changes

- Updated dependencies [89f1c423]
  - @sap-cloud-sdk/util@2.5.0
  - @sap-cloud-sdk/connectivity@2.5.0

## 2.4.0

### Patch Changes

- Updated dependencies [04726a35]
- Updated dependencies [0a008674]
- Updated dependencies [0a008674]
- Updated dependencies [0a008674]
  - @sap-cloud-sdk/connectivity@2.4.0
  - @sap-cloud-sdk/util@2.4.0
