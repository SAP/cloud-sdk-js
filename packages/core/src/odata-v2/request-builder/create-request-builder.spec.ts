import nock = require('nock');
import { v4 as uuid } from 'uuid';
import {
  defaultDestination,
  defaultHost,
  mockCreateRequest
} from '../../../test/test-util/request-mocker';
import { testEntityResourcePath } from '../../../test/test-util/test-data';
import {
  TestEntity,
  TestEntityMultiLink,
  TestEntitySingleLink
} from '../../../test/test-util/test-services/v2/test-service';
import { testPostRequestOutcome } from '../../../test/test-util/testPostRequestOutcome';
import { CreateRequestBuilder } from './create-request-builder';

describe('CreateRequestBuilder', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('create an entity with field properties', async () => {
    const keyProp = uuid();
    const stringProp = 'testStr';
    const postBody = { KeyPropertyGuid: keyProp, StringProperty: stringProp };

    mockCreateRequest({
      body: postBody,
      path: 'A_TestEntity'
    });

    const entity = TestEntity.builder()
      .keyPropertyGuid(keyProp)
      .stringProperty(stringProp)
      .build();

    const actual = await new CreateRequestBuilder(TestEntity, entity).execute(
      defaultDestination
    );

    testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
  });

  it('should set the remote state and ETag', async () => {
    const stringProp = 'etagTest';
    const eTag = 'someEtag';

    const postBody = {
      StringProperty: stringProp
    };

    mockCreateRequest({
      body: postBody,
      path: 'A_TestEntity',
      responseBody: {
        ...postBody,
        __metadata: { etag: eTag }
      }
    });

    const entity = TestEntity.builder().stringProperty(stringProp).build();
    const actual = await new CreateRequestBuilder(TestEntity, entity).execute(
      defaultDestination
    );

    expect(actual['_versionIdentifier']).toBe(eTag);
    expect(actual['remoteState']).toEqual(entity);
  });

  it('create an entity with a single link property', async () => {
    const stringProp = 'test';
    const postBody = { to_SingleLink: { StringProperty: stringProp } };

    mockCreateRequest({
      body: postBody,
      path: 'A_TestEntity'
    });

    const entity = TestEntity.builder()
      .toSingleLink(
        TestEntitySingleLink.builder().stringProperty(stringProp).build()
      )
      .build();

    const actual = await new CreateRequestBuilder(TestEntity, entity).execute(
      defaultDestination
    );

    testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
  });

  it('create an entity with a multi link property', async () => {
    const keyProp = 'test';
    const stringProp = 'someStr';

    const linkedEntityBody = [
      { KeyProperty: keyProp, StringProperty: stringProp }
    ];

    mockCreateRequest({
      body: { to_MultiLink: linkedEntityBody },
      path: 'A_TestEntity',
      responseBody: { to_MultiLink: { results: linkedEntityBody } }
    });

    const entity = TestEntity.builder()
      .toMultiLink([
        TestEntityMultiLink.builder()
          .keyProperty(keyProp)
          .stringProperty(stringProp)
          .build()
      ])
      .build();

    const actual = await new CreateRequestBuilder(TestEntity, entity).execute(
      defaultDestination
    );

    testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
  });

  it('create an entity with custom fields', async () => {
    const keyProp = uuid();
    const stringProp = 'test';

    const customFields = {
      CustomField1: 'abcd',
      CustomField2: 1234,
      CustomField3: { Attribute1: '1', Attribute2: false }
    };

    const postBody = {
      KeyPropertyGuid: keyProp,
      StringProperty: stringProp,
      ...customFields
    };

    mockCreateRequest({
      body: postBody,
      path: 'A_TestEntity'
    });

    const entity = TestEntity.builder()
      .keyPropertyGuid(keyProp)
      .stringProperty(stringProp)
      .withCustomFields(customFields)
      .build();

    const actual = await new CreateRequestBuilder(TestEntity, entity).execute(
      defaultDestination
    );

    testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
  });

  it('create entity as a child of an entity', async () => {
    const booleanProp = false;
    const int16Prop = 17;
    const parentKeyGuid = uuid();
    const parentKeyString = 'test-key';

    const childEntity = TestEntityMultiLink.builder()
      .booleanProperty(booleanProp)
      .int16Property(int16Prop)
      .build();

    const parentEntity = TestEntity.builder()
      .keyPropertyGuid(parentKeyGuid)
      .keyPropertyString(parentKeyString)
      .build();

    const postBody = { BooleanProperty: booleanProp, Int16Property: int16Prop };

    const toChildPath = `${testEntityResourcePath(
      parentKeyGuid,
      parentKeyString
    )}/to_MultiLink`;

    mockCreateRequest({
      body: postBody,
      path: toChildPath
    });

    const actual = await new CreateRequestBuilder(
      TestEntityMultiLink,
      childEntity
    )
      .asChildOf(parentEntity, TestEntity.TO_MULTI_LINK)
      .execute(defaultDestination);

    testPostRequestOutcome(actual, childEntity.setOrInitializeRemoteState());
  });

  it('throws an error when request execution fails', async () => {
    mockCreateRequest({
      body: () => true,
      statusCode: 500
    });

    const someEntity = TestEntity.builder().stringProperty('').build();

    const createRequest = new CreateRequestBuilder(
      TestEntity,
      someEntity
    ).execute(defaultDestination);

    await expect(createRequest).rejects.toThrowErrorMatchingSnapshot();
  });

  it('create an entity with csrf token request when the option is set to false', async () => {
    const keyProp = uuid();
    const stringProp = 'testStr';
    const postBody = { KeyPropertyGuid: keyProp, StringProperty: stringProp };

    nock(defaultHost)
      .post(
        '/testination/sap/opu/odata/sap/API_TEST_SRV/A_TestEntity',
        postBody
      )
      .reply(200, { d: postBody }, {});

    const entity = TestEntity.builder()
      .keyPropertyGuid(keyProp)
      .stringProperty(stringProp)
      .build();

    const actual = await new CreateRequestBuilder(TestEntity, entity)
      .skipCsrfTokenFetching()
      .execute(defaultDestination);

    testPostRequestOutcome(actual, entity.setOrInitializeRemoteState());
  });

  describe('executeRaw', () => {
    it('returns request and raw response', async () => {
      const keyProp = uuid();
      const stringProp = 'testStr';
      const postBody = { KeyPropertyGuid: keyProp, StringProperty: stringProp };

      mockCreateRequest({
        body: postBody,
        path: 'A_TestEntity'
      });

      const entity = TestEntity.builder()
        .keyPropertyGuid(keyProp)
        .stringProperty(stringProp)
        .build();

      const actual = await new CreateRequestBuilder(
        TestEntity,
        entity
      ).executeRaw(defaultDestination);

      expect(actual.data.d).toEqual(postBody);
      expect(actual.request.method).toBe('POST');
    });
  });
});
