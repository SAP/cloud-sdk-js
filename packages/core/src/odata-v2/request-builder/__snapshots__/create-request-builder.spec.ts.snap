// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRequestBuilder throws an error when request execution fails 1`] = `"Create request failed!"`;
